src/enable_variable_ops/ngraph_replace_optimizers.cc

```diff
@@ -0,0 +1,283 @@
+          continue;
+        }
+        // Check REF TYPE RATHER THAN NAME
+        if (edge->dst()->IsOp() && !edge->IsControlEdge() &&
+            IsRefType(edge->dst()->input_type(edge->dst_input()))) {
+          input_ref = NodeBuilder::NodeOut(edge->src(), edge->src_output());
+        } else {
+          input_val = NodeBuilder::NodeOut(edge->src(), edge->src_output());
+        }
+      }
+      NGRAPH_VLOG(1) << "Has assigned device " << node->assigned_device_name();
+
+      Node* sub_op;
+      string new_name_sub = node->name() + "_Sub";
+      TF_RETURN_IF_ERROR(NodeBuilder(new_name_sub, "Sub")
+                             .Input(input_ref)
+                             .Input(input_val)
+                             .Attr("T", dtype)
+                             .Device(node->assigned_device_name())
+                             .Finalize(graph, &(sub_op)));
+      sub_op->set_assigned_device_name(node->assigned_device_name());
+      NodeBuilder::NodeOut ndef_sub_op = NodeBuilder::NodeOut(sub_op, 0);
+
+      NGRAPH_VLOG(1) << "Sub op name: " << sub_op->name();
+      NGRAPH_VLOG(1) << "Sub op assigned device: "
+                     << sub_op->assigned_device_name();
+
+      Node* ngraphassign_op;
+      string new_name_ngassign = node->name() + "_NGraphAssign";
+
+      TF_RETURN_IF_ERROR(NodeBuilder(new_name_ngassign, "NGraphAssign")
+                             .Attr("validate_shape", true)
+                             .Attr("use_locking", true)
+                             .Attr("T", dtype)
+                             .Attr("ngraph_graph_id", 0)
+                             .Input(input_ref)
+                             .Input(ndef_sub_op)
+                             .Device(node->assigned_device_name())
+                             .Finalize(graph, &ngraphassign_op));
+      ngraphassign_op->set_assigned_device_name(node->assigned_device_name());
+      NGRAPH_VLOG(1) << "Assign op name: " << ngraphassign_op->name();
+      NGRAPH_VLOG(1) << "Assign op assigned device: "
+                     << ngraphassign_op->assigned_device_name();
+
+      for (auto edge : node->in_edges()) {
+        if (edge->IsControlEdge()) {
+          graph->AddEdge(edge->src(), edge->src_output(), ngraphassign_op,
+                         edge->dst_input());
+          graph->RemoveEdge(edge);
+        }
+      }
+      NGRAPH_VLOG(1) << "Added input control edges";
+      std::vector<const Edge*> edges;
+      for (auto edge : node->out_edges()) {
+        edges.push_back(edge);
+      }
+
+      NGRAPH_VLOG(1) << "Got output edges";
+
+      for (auto edge : edges) {
+        graph->AddEdge(ngraphassign_op, edge->src_output(), edge->dst(),
+                       edge->dst_input());
+        graph->RemoveEdge(edge);
+      }
+
+      NGRAPH_VLOG(1) << "Replaced output edges";
+
+      graph->RemoveNode(node);
+      NGRAPH_VLOG(1) << "Removing node";
+
+    }  // AssignSub
+    else if (node->type_string() == "NGraphAssignAdd") {
```

Comment:
Maybe the `NGraphAssignAdd` and `NGraphAssignSub` cases can be collapsed into one